index.rst

```diff
@@ -10,20 +10,357 @@
+
+``Butler.import`` and ``Butler.transfer_from``
+""""""""""""""""""""""""""""""""""""""""""""""
+
+As with ``put``, we would write a journal file before the Datastore operations begin and delete it after Registry writes succeed.
+
+``Butler.get``
+""""""""""""""
+
+No journaling is needed, as this is a read-only operation.
+
+``Butler.unstore`` and other removals
+"""""""""""""""""""""""""""""""""""""
+
+The journal file should be written before the ``Registry`` transaction is committed and deleted only after all Datastore deletions succeed.
+This is slightly problematic for client/server, because the journal file will need to be populated with information we get from the Registry database; this means the client cannot be responsible for creating the journal file unless we make fetching the datastore records and deleting them separate operations.
+That isn't too bad - it's just a slight increase in latency and a bit more http traffic.
+Another alternative would be to have the server take responsibility for creating the journal file, and then either returning responsibility for its deletion to the client or taking responsibility for both the deletion of the Datastore artifacts and the deletion of the journal file.
+Which of these is preferable probably depends on whether we want these operations to block until completion and whether we have reasons to perform other Datastore operations on the server (up to this point, having the client use a vanilla ``FileDatastore`` and perform all Datastore operations still seems viable).
+
+.. note::
+   This proposal should not affect our ability support disassembled composites, though we may be able to make further simplifications if we drop that support.
+   It may have implications for our ability to support multiple datasets in a single file, at least in terms of safeguarding against premature deletions
+   of those files.
+   An easy way to mitigate that would be to limit that support to "unmanaged" datasets that are ingested with absolute URIs, since those are never deleted by butler code, and this satisfies Rubin's only current use case (ingesting DECam raws) for this functionality.
+
+.. _including-signed-urls-for-access-control:
+
+Including signed URLs for access control
+----------------------------------------
+
+Our access control model for the official Rubin Observatory data repositories (see :cite:`DMTN-169`) is based on information stored in the Registry - collection names, whenever possible, and new naming conventions or new columns in contexts (e.g. dimension records or dataset type names) that are not associated with collections.
+Access to the associated files managed by a Datastore is mediated by signed URLs; once the server side of the remote Butler has determined that an API call is permitted (based on that the Registry-side information), it can generate one or more signed URLs to pass to Datastore that provide direct access to the controlled files.
```

Comment:
If we need to go through Gafaelfawr to sign URLs, I think that's ok, as long as we can do it in bulk rather than one at a time.  But I think the Butler server will have to contact Gafaelfawr up front to authenticate each request and obtain some sort of summary of what that user can access, and from that the Butler server should have everything it needs to know _whether_ to sign a URL (or even perform a query against a particular collection) without contacting Gafaelfawr again.

Rate-limiting between Butler server and whatever service it calls to do the signing sounds like it might make having an async server implementation even more important, but otherwise I don't see an obvious problem with it.

I'm pretty out of my depth on this part, though, so feel free to instead tell me how you think it ought to be done and I can tell you if that's problematic from my end :slightly_smiling_face: .